### PR TITLE
Feature/single server smart graph dump restore tests

### PR DESCRIPTION
**Pull request by adityamukho**
### Scope & Purpose

Adds mixed dump-restore tests for smart graphs. Dump from single server, restore on cluster.

What is covered:
1. Smart Graphs

What is pending:
1. Disjoint Smart Graphs
2. Satellite Graphs

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR: https://github.com/arangodb/enterprise/pull/767
- [ ] GitHub issue / Jira ticket number: https://arangodb.atlassian.net/browse/GORDO-1304
- [ ] Design document: https://docs.google.com/document/d/1wpjd3lNA-Voal4-oCPgEMZ6JlXN8MM54xdxDJhj3ugE/edit

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: tests/js/server/dump/dump-setup-common.inc [hunk 1]
```diff
@@ -442,7 +465,67 @@ function createSmartGraph() {
...
+  const orphans = "UnitTestDumpSmartOrphansSingleServer";
+  if (sgm._exists(smartGraphName)) {
+    sgm._drop(smartGraphName, true);
+  }
+  db._drop(edges);
+  db._drop(vertices);
+  db._drop(orphans);
+
+  sgm._create(smartGraphName, [sgm._relation(edges, vertices, vertices)],
+    [orphans], {numberOfShards: 5, smartGraphAttribute: "value"});
```
Comment (romanatarango) [inline tests/js/server/dump/dump-setup-common.inc:488]: This part could be replaced with `createSmartGraphWithoutData()` with appropriate parameters for the names.

File: tests/js/server/dump/dump-test.inc [hunk 2]
```diff
@@ -1184,7 +1185,136 @@
...
       },
-      
+
+      ////////////////////////////////////////////////////////////////////////////////
+      /// @brief test smart graph restores
+      ////////////////////////////////////////////////////////////////////////////////
+      testEmptySmartGraph: function () {
+        const smartGraphName = "UnitTestsDumpEmptySmartGraph";
+        const g = db._collection('_graphs');
+        console.warn(g.toArray());
```
Comment (romanatarango) [inline tests/js/server/dump/dump-test.inc:1195]: This is not needed any more.

File: tests/js/server/dump/dump-test.inc [hunk 3]
```diff
@@ -1184,7 +1185,136 @@
...
         }
       },
-      
+
+      ////////////////////////////////////////////////////////////////////////////////
+      /// @brief test smart graph restores
+      ////////////////////////////////////////////////////////////////////////////////
+      testEmptySmartGraph: function () {
+        const smartGraphName = "UnitTestsDumpEmptySmartGraph";
+        const g = db._collection('_graphs');
```
Comment (romanatarango) [inline tests/js/server/dump/dump-test.inc:1194]: `const graphs` may be clearer

File: tests/js/server/dump/dump-test.inc [hunk 4]
```diff
@@ -1184,7 +1185,136 @@
           assertTrue(next < 100000000, "expected next uniqid after restore to be less than 100000000, not " + next);
         }
       },
-      
+
+      ////////////////////////////////////////////////////////////////////////////////
+      /// @brief test smart graph restores
+      ////////////////////////////////////////////////////////////////////////////////
+      testEmptySmartGraph: function () {
+        const smartGraphName = "UnitTestsDumpEmptySmartGraph";
```
Comment (romanatarango) [inline tests/js/server/dump/dump-test.inc:1193]: The name is already defined in `tests/js/server/dump/dump-setup-common.inc` and should be passed here as a parameter.

File: tests/js/server/dump/dump-test.inc [hunk 5]
```diff
@@ -1184,7 +1185,136 @@
...
+        assertEqual(graph.smartGraphAttribute, 'value');
+        assertEqual(graph.edgeDefinitions.length, 0);
+        assertEqual(graph.orphanCollections.length, 0);
+      },
+
+      testSmartGraphWithoutData: function() {
+        const smartGraphName = "UnitTestDumpSmartGraphWithoutData";
+        const vertexCollName = 'UnitTestDumpSmartVerticesWithoutData';
+        const edgeCollName = 'UnitTestDumpSmartEdgesWithoutData';
+        const orphanCollName = 'UnitTestDumpSmartOrphansWithoutData';
```
Comment (romanatarango) [inline tests/js/server/dump/dump-test.inc:1213]: The names should be passed as parameters.

File: tests/js/server/dump/dump-test.inc [hunk 6]
```diff
@@ -1184,7 +1185,136 @@
...
+        const g = db._collection('_graphs');
+        assertTrue(!!g.exists(smartGraphName));
+
+        const graph = g.document(smartGraphName);
+        assertFalse(graph.isDisjoint);
+        assertFalse(graph.isSatellite);
+        assertTrue(graph.isSmart);
+        assertEqual(graph.replicationFactor, 1);
+        assertEqual(graph.numberOfShards, 5);
+        assertEqual(graph.smartGraphAttribute, 'value');
```
Comment (romanatarango) [inline tests/js/server/dump/dump-test.inc:1224]: These asserts are the same multiple times, this could be done once in a separate function.

**Review by romanatarango — COMMENTED** [review]
It may be interesting also to test non-default replication factor, number of shards and sharding strategy to know that this information is not lost when dumping in single server and restoring in a cluster.

File: tests/js/server/dump/dump-test.inc (hunk 4 above)
Comment (adityamukho) [reply tests/js/server/dump/dump-test.inc:1193]: This and other tests added in the PR are special cases that benefit from qualifiers in their name, also to keep them unique.

The pre-defined name would not be suitable here.

File: tests/js/server/dump/dump-test.inc (hunk 6 above)
Comment (adityamukho) [reply tests/js/server/dump/dump-test.inc]: Fixed.

File: tests/js/server/dump/dump-test.inc (hunk 3 above)
Comment (adityamukho) [reply tests/js/server/dump/dump-test.inc]: Fixed.

File: tests/js/server/dump/dump-test.inc (hunk 2 above)
Comment (adityamukho) [reply tests/js/server/dump/dump-test.inc]: Fixed.

File: tests/js/server/dump/dump-setup-common.inc (hunk 1 above)
Comment (adityamukho) [reply tests/js/server/dump/dump-setup-common.inc]: Fixed.

File: tests/js/server/dump/dump-test.inc (hunk 5 above)
Comment (adityamukho) [reply tests/js/server/dump/dump-test.inc:1248]: I don't think test cases accept parameters.

**Review by romanatarango — APPROVED** [review]

**Pull request merged**